src/carma-1-tenth/plot_absolute_route_deviation.py

```diff
@@ -0,0 +1,103 @@
+from rosbag_utils import open_bagfile
+import numpy as np
+import yaml
+import tqdm
+from rosidl_runtime_py.utilities import get_message
+from rclpy.serialization import deserialize_message
+import datetime as dt
+import matplotlib.pyplot as plt
+from scipy.interpolate import make_interp_spline
+import os
+
+
+def find_closest_point(point_arr, point):
+    difference_arr = np.linalg.norm(point_arr - point, axis=1)
+    min_index = difference_arr.argmin()
+    # Don't want to include deviations if we have not yet reached the route or have completed it
+    if min_index == 0 or min_index == len(difference_arr) - 1:
+        return None
+    return point_arr[min_index]
+
+
+def plot_absolute_route_deviation(bag_dir, start_offset=0.0):
+    metadatafile : str = os.path.join(bag_dir, "metadata.yaml")
+    if not os.path.isfile(metadatafile):
+        raise ValueError("Metadata file %s does not exist. Are you sure %s is a valid rosbag?" % (metadatafile, bag_dir))
+    with open(metadatafile, "r") as f:
+        metadata_dict : dict = yaml.load(f, Loader=yaml.SafeLoader)["rosbag2_bagfile_information"]
+    storage_id = metadata_dict['storage_identifier']
+    odom_topic = '/amcl_pose'
+    route_topic = '/route_graph'
+    reader, type_map = open_bagfile(bag_dir, topics=[odom_topic, route_topic], storage_id=storage_id)
+    topic_count_dict = {entry["topic_metadata"]["name"] : entry["message_count"] for entry in metadata_dict["topics_with_message_count"]}
+
+    route_graph = None
+    odom_count = 0
+    odometry = np.zeros((topic_count_dict[odom_topic], 2))
+    odometry_times = np.zeros((topic_count_dict[odom_topic],))
+    # Iterate through bag and store odometry + route_graph messages
+    for idx in tqdm.tqdm(iterable=range(topic_count_dict[odom_topic] + topic_count_dict[route_topic])):
+        if(reader.has_next()):
+            (topic, data, t_) = reader.read_next()
+            msg_type = type_map[topic]
+            msg_type_full = get_message(msg_type)
+            msg = deserialize_message(data, msg_type_full)
+            if msg_type == "visualization_msgs/msg/MarkerArray":
```

Comment:
Can we replace this with the topic name as well?

Reply:
done!